### PR TITLE
Implement rule status for word break boundary.

### PR DESCRIPTION
**Pull request by makotokato**
Fixes https://github.com/unicode-org/icu4x/issues/2755

### CODE REVIEW

**Review by sffc — DISMISSED** [review]
Code looks good, but I defer on the logical correctness

File: experimental/segmenter/src/provider.rs [hunk 1]
```diff
@@ -84,6 +88,18 @@ pub struct RuleBreakStateTable<'data>(
     #[cfg_attr(feature = "serde", serde(borrow))] pub ZeroVec<'data, i8>,
 );
 
+/// Rules status data
```
Comment (Manishearth) [inline experimental/segmenter/src/provider.rs:91]: nit: please specify how this works in the comment

File: experimental/segmenter/src/rule_segmenter.rs [hunk 2]
```diff
@@ -9,6 +9,18 @@ use crate::symbols::*;
 use core::str::CharIndices;
 use utf8_iter::Utf8CharIndices;
 
+/// The category tag that is returned by rule_status.
+#[non_exhaustive]
+#[derive(PartialEq, Debug)]
+pub enum RuleStatusType {
+    /// No category tag
+    None = 0,
```
Comment (robertbastian) [inline experimental/segmenter/src/rule_segmenter.rs:17]: I think you need `#[repr(u8)]`, as you cast to `u8` in datagen.
Consider storing the enum directly in the `ZeroVec`, so you don't have to do the conversions.

File: experimental/segmenter/src/rule_segmenter.rs [hunk 3]
```diff
@@ -216,6 +238,39 @@ impl<'l, 's, Y: RuleBreakType<'l, 's>> RuleBreakIterator<'l, 's, Y> {
...
+        {
+            match value {
+                1 => RuleStatusType::Number,
+                2 => RuleStatusType::Letter,
+                _ => RuleStatusType::None,
+            }
+        } else {
+            RuleStatusType::None
+        }
+    }
```
Comment (robertbastian) [inline experimental/segmenter/src/rule_segmenter.rs:267]: nit: could be simplified

```suggestion
        match self
            .data
            .rule_status_table
            .0
            .get((self.boundary_property - 1) as usize)
        {
            Some(1) => RuleStatusType::Number,
            Some(2) => RuleStatusType::Letter,
            _ => RuleStatusType::None,
        }
    }
```

File: experimental/segmenter/src/rule_segmenter.rs [hunk 4]
```diff
@@ -216,6 +238,39 @@ impl<'l, 's, Y: RuleBreakType<'l, 's>> RuleBreakIterator<'l, 's, Y> {
...
+        if let Some(value) = self
+            .data
+            .rule_status_table
+            .0
+            .get((self.boundary_property - 1) as usize)
+        {
+            match value {
+                1 => RuleStatusType::Number,
+                2 => RuleStatusType::Letter,
+                _ => RuleStatusType::None,
```
Comment (robertbastian) [inline experimental/segmenter/src/rule_segmenter.rs:262]: Consider encoding the `None` case as `Option<RuleStatusType>`

File: ffi/diplomat/tests/missing_apis.txt [hunk 5]
```diff
@@ -123,3 +123,4 @@ icu::relativetime::provider::ShortWeekRelativeTimeFormatDataV1Marker#Struct
 icu::relativetime::provider::ShortYearRelativeTimeFormatDataV1Marker#Struct
 icu::relativetime::provider::SingularSubPattern#Struct
 icu::relativetime::provider::SingularSubPattern::try_from_str#FnInStruct
+icu::segmenter::RuleStatusType#Enum
```
Comment (aethanyc) [inline ffi/diplomat/tests/missing_apis.txt:126]: suggestion: we probably want to expose `is_word_like()` and `rule_status` in FFI (e.g. in `segmenter_word.rs`). They can be added in a separate PR if you like.

File: experimental/segmenter/src/rule_segmenter.rs (hunk 2 above)
Comment (Manishearth) [reply experimental/segmenter/src/rule_segmenter.rs:17]: you don't need to set the repr for the cast to work

Comment (makotokato) [reply experimental/segmenter/src/rule_segmenter.rs:18]: Although new patch adds `#[repr(u8)]`, should I remove it? Even if adding it, we cannot remove cast.

File: ffi/diplomat/tests/missing_apis.txt (hunk 5 above)
Comment (makotokato) [reply ffi/diplomat/tests/missing_apis.txt:28]: After fixing this, I will create new PR

**Pull request merged**